src/uproot/source/fsspec.py

```diff
@@ -97,17 +97,21 @@ def chunks(self, ranges, notifications):
         self._num_requests += 1
         self._num_requested_chunks += len(ranges)
         self._num_requested_bytes += sum(stop - start for start, stop in ranges)
-        data = self._fs.cat_ranges(
-            [self._file_path] * len(ranges),
-            [start for start, _ in ranges],
-            [stop for _, stop in ranges],
-        )
+
+        executor = uproot.source.futures.ResourceThreadPoolExecutor(ranges)
```

Comment:
This executor goes out of scope, possibly/probably before the work is finished. I think that would shut it down prematurely (the work never finishes, which would cause the event loop in src/uproot/behaviors/TBranch.py to hang).

The HTTPSource and XRootDSource do this correctly: they have a `ResourceThreadPoolExecutor` attached to the Source (`self`). The "resource" is a file handle, something that absolutely must be closed when the remote file leaves its `with` block. Most kinds of file handles are limited in number (too many open file handles and the OS will complain).

`ranges` is not a limited resource. If the fsspec interface is truly stateless (no handle), then we can get away with a `ThreadPoolExecutor`, rather than a `ResourceThreadPoolExecutor`.

Reply:
My initial thought was also that the executor going out of scope prematurely would cause it to shutdown and not finish the work but it passes the test so it's somehow working.

Reply:
By accident, perhaps? I don't know. But we definitely can't rely on it.